### PR TITLE
Add Cluster properties and rename DeploymentId to ClusterId

### PR DESCRIPTION
**Pull request by nehmebilal**
This PR addresses issue #33 

For a quick overview of the changes in the Api and `DeploymentConfig.json`, take a look at the [tests](https://github.com/nehmebilal/Yams/commit/cbf47e0e03464dd8b0234a597cfcd415868dc071).

The new `DeploymentConfig.json` is as follows:

```json
{
    "Applications":
    [
        {
            "Id": "app1",
            "Version": "1.0.0",
            "TargetClusters": [ "YAMS_CLUSTER_ID" ],
            "Properties":
            {
                "NodeType": "PROD"
            }
        },  
        {
            "Id": "app1",
            "Version": "1.0.1",
            "TargetClusters": [ "YAMS_CLUSTER_ID"],
            "Properties":
            {
                "NodeType": "QA"
            }
        },
    ]
}
```

`DeploymentIds` has been renamed to `TargetClusters` and the `DeploymentId` token has been renamed to `ClusterId`. I didn't end up adding the 'Yams' prefix because it was quickly spreading through the code and felt heavy. If that's a major concern we can rename it.

Yams will match on properties by default (no need to create a custom matcher). To add properties to the cluster, you can use the `YamsConfigBuilder.AddClusterProperty("Foo", "Bar")` Api as shown in the tests. Only the properties that are added through YamsConfigBuilder will be used by the default matcher (other properties in the `DeploymentConfig.json` will be carried over but not used for matching). All properties can be used as token in the `AppConfig.json` (`${propName}`).

I also updated the sample and docs but not all of them. I'm hoping that Orleans guys (@jdom) will be able to help with the Orleans sample and cloud service sample.

Note that some of the Api changes are not backward compatible but should be easy to fix.

Finally, I extracted the `DeploymentConfig` serialization into its own interface so that it can be replaced. This means that you can now easily replace the json representation with whatever you prefer.

Let me know what you think!

### CODE REVIEW

File: Docs/Deploy&Host_an_App_in_YAMS.md [hunk 1]
```diff
@@ -234,12 +234,12 @@ To deploy the new version (let's say `2.0.0`) side-by-side with the currently ru
...
 			"Id": "WebApp",
 			"Version": "1.1.0",
-			"DeploymentIds": [ "MY_DEPLOYMENT_ID" ]
+			"TargetClusters": [ "MY_DEPLOYMENT_ID" ]
 		},
 		{
 			"Id": "WebApp",
 			"Version": "2.0.0",
-			"DeploymentIds": [ "MY_DEPLOYMENT_ID" ]
+			"TargetClusters": [ "MY_DEPLOYMENT_ID" ]
```
Comment (jdom) [inline Docs/Deploy&Host_an_App_in_YAMS.md]: should you update these to MY_CLUSTER_ID?

File: src/Etg.Yams.Core/Storage/Config/DeploymentConfig.cs [hunk 2]
```diff
@@ -1,305 +1,184 @@
...
+            if (ReferenceEquals(this, obj)) return true;
+            if (obj.GetType() != this.GetType()) return false;
+            return Equals((DeploymentConfig) obj);
         }
 
-        public IEnumerable<string> ListApplications()
+        public override int GetHashCode()
         {
-            return _apps.Keys;
+            return (_apps != null ? _apps.GetHashCode() : 0);
```
Comment (jdom) [inline src/Etg.Yams.Core/Storage/Config/DeploymentConfig.cs]: not sure for what purpose this relevant to override, but are you certain that 2 different sets with the same values will return the same hashcode? or is that not what you want anyway?

File: src/Etg.Yams/Etg.Yams.nuspec [hunk 3]
```diff
@@ -2,7 +2,7 @@
 <package >
   <metadata>
     <id>Etg.Yams</id>
-    <version>1.0.14</version>
+    <version>1.1.0-test</version>
```
Comment (jdom) [inline src/Etg.Yams/Etg.Yams.nuspec]: there's a lot of references to 1.1.0-test, is that something that you want merged?

File: test/Etg.Yams.Core.Test/Storage/DeploymentConfigTest.cs [hunk 4]
```diff
@@ -2,36 +2,40 @@
...
 using Etg.Yams.Storage.Config;
 using Etg.Yams.Test.Utils;
+using Etg.Yams.Utils;
 using Xunit;
 using Newtonsoft.Json.Linq;
+using Newtonsoft.Json.Serialization;
 
 namespace Etg.Yams.Test.Storage
 {
     public class DeploymentConfigTestFixture
```
Comment (jdom) [inline test/Etg.Yams.Core.Test/Storage/DeploymentConfigTest.cs]: might want to add coverage for using the new `AddApplication` overload that takes a deployment config with properties

File: Docs/Deploy&Host_an_App_in_YAMS.md (hunk 1 above)
Comment (nehmebilal) [inline Docs/Deploy&Host_an_App_in_YAMS.md]: Yep thanks

File: src/Etg.Yams.Core/Storage/Config/DeploymentConfig.cs (hunk 2 above)
Comment (nehmebilal) [inline src/Etg.Yams.Core/Storage/Config/DeploymentConfig.cs]: I'll test this a bit more. I generated it the implementation with Resharper. I'd expect that two sets with the same values would have the same hashcode (GetHashCode is supposed to be consistent with Equals) but I'll make sure.

File: src/Etg.Yams/Etg.Yams.nuspec (hunk 3 above)
Comment (nehmebilal) [inline src/Etg.Yams/Etg.Yams.nuspec]: No they should be replaced with the actual nuget version that will be published. I'll clean that up before merging, thanks.

File: test/Etg.Yams.Core.Test/Storage/DeploymentConfigTest.cs (hunk 4 above)
Comment (nehmebilal) [inline test/Etg.Yams.Core.Test/Storage/DeploymentConfigTest.cs]: Good point!

File: src/Etg.Yams.Core/Storage/Config/DeploymentConfig.cs (hunk 2 above)
Comment (veikkoeeva) [inline src/Etg.Yams.Core/Storage/Config/DeploymentConfig.cs]: Possible could be also `EqualityComparer<HashSet<AppDeploymentConfig>.Default.GetHashCode(_apps)`, I think.

File: Docs/Deploy&Host_an_App_in_YAMS.md (hunk 1 above)
Comment (nehmebilal) [inline Docs/Deploy&Host_an_App_in_YAMS.md]: Done.

File: src/Etg.Yams.Core/Storage/Config/DeploymentConfig.cs (hunk 2 above)
Comment (nehmebilal) [inline src/Etg.Yams.Core/Storage/Config/DeploymentConfig.cs]: @jdom was right, the default implementation doesn't handle IEnumerable's correctly. I fixed the implementation and added corresponding tests.

File: src/Etg.Yams/Etg.Yams.nuspec (hunk 3 above)
Comment (nehmebilal) [inline src/Etg.Yams/Etg.Yams.nuspec]: Done.

File: Docs/Deploy&Host_an_App_in_YAMS.md (hunk 1 above)
Comment (nehmebilal) [inline Docs/Deploy&Host_an_App_in_YAMS.md]: Done

File: src/Etg.Yams.Core/Storage/Config/DeploymentConfig.cs (hunk 2 above)
Comment (nehmebilal) [inline src/Etg.Yams.Core/Storage/Config/DeploymentConfig.cs]: @jdom was right, the default implementation doesn't handle IEnumerable's correctly. I fixed the implementation and added some tests to confirm.

File: src/Etg.Yams/Etg.Yams.nuspec (hunk 3 above)
Comment (nehmebilal) [inline src/Etg.Yams/Etg.Yams.nuspec]: Replaced with 1.1.0-rc

File: test/Etg.Yams.Core.Test/Storage/DeploymentConfigTest.cs (hunk 4 above)
Comment (nehmebilal) [inline test/Etg.Yams.Core.Test/Storage/DeploymentConfigTest.cs]: Tests added. Found some issues and ended up replacing the internal HashSet with a Dictionary which was more appropriate.

**Comment by nehmebilal** [comment]
@jdom Please confirm if your comments have been addressed and then I'll merge and publish a NuGet.

**Comment by jdom** [comment]
LGTM, ship it :P

**Comment by nehmebilal** [comment]
One last change. I made the cluster properties also available as tokens (properties added to the cluster at startup) and fixed a bug in resolving version tokens that was introduced by the switch to SemVersion (version.Patch should be used instead of version.Build).

**Pull request merged**